### PR TITLE
Improve dump option support

### PR DESCRIPTION
**Pull request by chqrlie**
- add `JS_SetDumpFlags()` to select active dump options
- accept -d[<hex mask>] and --dump[=<hex mask>] to specify active dump options, generalize command line option handling
- make all defined `DUMP_XXX` output dynamically selectable
- improve `DUMP_READ_OBJECT` output, fix indentation issue

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -170,6 +170,7 @@ target_compile_definitions(qjs PRIVATE ${qjs_defines})
 if (CMAKE_BUILD_TYPE MATCHES Debug OR DUMP_LEAKS)
     target_compile_definitions(qjs PRIVATE
         DUMP_LEAKS
+        DUMP_LEAKS_FLAG=0
```
Comment (saghul) [inline CMakeLists.txt]: Wouldn't this disable it?

File: qjs.c [hunk 2]
```diff
@@ -384,20 +395,28 @@ int main(int argc, char **argv)
...
+                        fprintf(stderr, "expecting memory limit");
+                        exit(1);
+                    }
+                    opt_arg = argv[optind++];
                 }
-                memory_limit = (size_t)strtod(argv[optind++], NULL);
-                continue;
+                // TODO(chqrlie): accept kmg suffixes
+                memory_limit = (size_t)strtod(opt_arg, NULL);
+                break;
```
Comment (saghul) [inline qjs.c:407]: Shouldn't we continue here?

File: qjs.c [hunk 3]
```diff
@@ -384,20 +395,28 @@ int main(int argc, char **argv)
...
+                        fprintf(stderr, "expecting stack size");
+                        exit(1);
+                    }
+                    opt_arg = argv[optind++];
                 }
-                stack_size = (size_t)strtod(argv[optind++], NULL);
-                continue;
+                // TODO(chqrlie): accept kmg suffixes
+                stack_size = (size_t)strtod(opt_arg, NULL);
+                break;
```
Comment (saghul) [inline qjs.c:419]: Ditto

File: quickjs.h [hunk 4]
```diff
@@ -294,6 +294,7 @@ JS_EXTERN JSRuntime *JS_NewRuntime(void);
 /* info lifetime must exceed that of rt */
 JS_EXTERN void JS_SetRuntimeInfo(JSRuntime *rt, const char *info);
 JS_EXTERN void JS_SetMemoryLimit(JSRuntime *rt, size_t limit);
+JS_EXTERN void JS_SetDumpFlags(JSRuntime *rt, unsigned flags);
```
Comment (saghul) [inline quickjs.h]: If we are making the flags configurable, how about making the stream where we dump the output configurable too?

File: quickjs.c [hunk 5]
```diff
@@ -72,37 +72,49 @@
...
+//#define DUMP_BYTECODE_FLAG_PASS3   0x01   /* 1: dump pass 3 final byte code */
+//#define DUMP_BYTECODE_FLAG_PASS2   0x02   /* 2: dump pass 2 code */
+//#define DUMP_BYTECODE_FLAG_PASS1   0x04   /* 4: dump pass 1 code */
+//#define DUMP_BYTECODE_FLAG_STDLIB  0x08   /* 8: dump stdlib functions */
+//#define DUMP_BYTECODE_FLAG_HEX     0x10   /* 16: dump bytecode in hex */
+//#define DUMP_BYTECODE_FLAG_PC2LINE 0x20   /* 32: dump line number table */
+//#define DUMP_BYTECODE_FLAG_STACK   0x40   /* 64: dump compute_stack_size */
+//#define DUMP_BYTECODE_FLAG_STEP    0x80   /* 128: dump executed bytecode */
+
+// dump options selected dynamically with -d<hex mask>.
```
Comment (saghul) [inline quickjs.c:86]: This is a bit weird. If we have an API to enable dumping but one doesn't uncomment the defined it doesn't really work, which is somewhat unexpected.

Perhaps we could have some if-def for all of them at once, and then you can select which one you want with the flags? We could set that one on Debug builds for example. Or would the overhead be too much?

File: quickjs.c [hunk 6]
```diff
@@ -72,37 +72,49 @@
...
+//#define FORCE_GC_AT_MALLOC  /* test the GC by forcing it before each object allocation */
+//#define DUMP_GC_FREE    /* dump objects freed by the GC */
+//#define DUMP_GC_FREE_FLAG        0x800
+
+//#define DUMP_MODULE_RESOLVE  /* dump module resolution steps */
+//#define DUMP_MODULE_RESOLVE_FLAG   0x1000
+//#define DUMP_PROMISE  /* dump promise steps */
+//#define DUMP_PROMISE_FLAG        0x2000
+
+//#define DUMP_LEAKS  2   /* dump leaked objects in JS_FreeContext (2: include atoms) */
```
Comment (saghul) [inline quickjs.c]: Does this define require a value?

File: quickjs.c [hunk 7]
```diff
@@ -1029,19 +1043,15 @@ static __exception int JS_ToArrayLengthFree(JSContext *ctx, uint32_t *plen,
...
 JSValue __attribute__((format(printf, 2, 3))) JS_ThrowInternalError(JSContext *ctx, const char *fmt, ...);
-static __maybe_unused void JS_DumpAtoms(JSRuntime *rt);
-static __maybe_unused void JS_DumpString(JSRuntime *rt,
-                                                  const JSString *p);
+
+static __maybe_unused void JS_DumpString(JSRuntime *rt, const JSString *p);
 static __maybe_unused void JS_DumpObjectHeader(JSRuntime *rt);
 static __maybe_unused void JS_DumpObject(JSRuntime *rt, JSObject *p);
 static __maybe_unused void JS_DumpGCObject(JSRuntime *rt, JSGCObjectHeader *p);
-static __maybe_unused void JS_DumpValueShort(JSRuntime *rt,
```
Comment (saghul) [inline quickjs.c]: Looks like this could be a separate PR?

File: quickjs.h (hunk 4 above)
Comment (saghul) [inline quickjs.h]:
```suggestion
JS_EXTERN void JS_SetDumpFlags(JSRuntime *rt, uint32_t flags);
```

Since that's how we define the field internally? I'd also make it uint64_t, since we might add more in the future...

File: quickjs.c [hunk 8]
```diff
@@ -1927,7 +1944,7 @@ void JS_FreeRuntime(JSRuntime *rt)
 
 #ifdef DUMP_LEAKS
     /* leaking objects */
-    {
+    if ((rt->dump_flags & DUMP_LEAKS_FLAG) == DUMP_LEAKS_FLAG) {
```
Comment (saghul) [inline quickjs.c]: All of these ones that only really affect FreeRuntime could be without an ifdef IMHO since there won't be a bug penalty...

File: quickjs.c [hunk 9]
```diff
@@ -11730,7 +11755,11 @@ static __maybe_unused void JS_DumpValueShort(JSRuntime *rt,
         {
             JSFunctionBytecode *b = JS_VALUE_GET_PTR(val);
             char buf[ATOM_GET_STR_BUF_SIZE];
-            printf("[bytecode %s]", JS_AtomGetStrRT(rt, buf, sizeof(buf), b->func_name));
+            if (b->func_name) {
```
Comment (saghul) [inline quickjs.c]: This could be a standalone change in a separate PR too...

File: quickjs.c [hunk 10]
```diff
@@ -14538,7 +14552,8 @@ static JSValue JS_CallInternal(JSContext *caller_ctx, JSValue func_obj,
     JSInlineCache *ic;
 
 #if defined(DUMP_BYTECODE) && (DUMP_BYTECODE & 128)
-#define DUMP_BYTECODE_OR_DONT(pc) dump_single_byte_code(ctx, pc, b);
+#define DUMP_BYTECODE_OR_DONT(pc) \
+    if (ctx->rt->dump_flags & 128) dump_single_byte_code(ctx, pc, b);
```
Comment (saghul) [inline quickjs.c]: Why not define this 128 as a flag too?

File: quickjs.c [hunk 11]
```diff
@@ -33922,13 +33968,23 @@ static JSValue JS_ReadFunctionTag(BCReaderState *s)
             vd->is_lexical = bc_get_flags(v8, &idx, 1);
             vd->is_captured = bc_get_flags(v8, &idx, 1);
 #ifdef DUMP_READ_OBJECT
-            bc_read_trace(s, "name: "); print_atom(s->ctx, vd->var_name); printf("\n");
+            if ((s->ctx->rt->dump_flags & DUMP_READ_OBJECT_FLAG) == DUMP_READ_OBJECT_FLAG) {
+                bc_read_trace(s, "%3d  %d%c%c%c %4d  ",
```
Comment (saghul) [inline quickjs.c]: Nice!

**Review by saghul — COMMENTED** [review]
Left some comments!

File: CMakeLists.txt (hunk 1 above)
Comment (chqrlie) [reply CMakeLists.txt]: setting `DUMP_LEAKS_FLAG` to `0` makes the output unconditional. Any other value makes the output selectable with the `-d<bitmask>` command line option (or a future javascript extended command).

File: qjs.c (hunk 2 above)
Comment (chqrlie) [reply qjs.c:407]: No, we should `break` from this loop if `opt_arg` was consumed so the next argument is scanned for options.  This is somewhat tricky, I will try and find a more intuitive approach, but the logic was already there and used incorrectly in this case.

File: qjs.c (hunk 3 above)
Comment (chqrlie) [reply qjs.c:419]: Same logic.

File: quickjs.h (hunk 4 above)
Comment (chqrlie) [reply quickjs.h]: Good point, I have another patch in the pipeline where the output stream is passed and used for the dump code, along with a custom version of printf to output `JSAtom`, `JSString` and JSValue` directly.

I made separate patches to keep them manageable.

Setting the output stream should be a different API.

File: quickjs.c (hunk 5 above)
Comment (chqrlie) [reply quickjs.c:86]: Actually, these defines are not used ATM for `DUMP_BYTECODE`, which is always dynamically selected. The `DUMP_XXX_FLAG` could be defined all the time or if not already defined using ifdefs if the `DUMP_XXX` is defined.  Work in progress I guess. 
I shall post a commit to make the output unconditional if the `DUMP_XXX_FLAG` is not defined already to simplify debug builds but I want to have a simple way to make the output conditional for interactive testing.

File: quickjs.h (hunk 4 above)
Comment (chqrlie) [reply quickjs.h]: The current implementation uses 24 bits, but we can make the API more futureproof.

Comment (saghul) [reply quickjs.h]: Sounds good!

File: quickjs.c (hunk 8 above)
Comment (chqrlie) [reply quickjs.c]: The `ifdef` is used to avoid compiling and linking the dump code if not required du reduce the library size.  This is useful for embedded targets and __wasi__.

Comment (chqrlie) [reply quickjs.c]: I am shall commit a simpler version too.

Comment (saghul) [reply quickjs.c]: Maybe some ENABLE_DUMPING (or a better name) which gates them all would be good?

File: quickjs.c (hunk 10 above)
Comment (chqrlie) [reply quickjs.c]: Yes, for completeness, the test should be:
```
#if defined(DUMP_BYTECODE) && (DUMP_BYTECODE & DUMP_BYTECODE_FLAG_STEP) == DUMP_BYTECODE_FLAG_STEP
#define DUMP_BYTECODE_OR_DONT(pc) \
    if (check_dump_flag(ctx->rt, DUMP_BYTECODE_FLAG_STEP)) ...
```

File: quickjs.c (hunk 11 above)
Comment (chqrlie) [reply quickjs.c]: I shall further simplify this in the next phase where `bc_read_trace` can output atoms directly.

File: quickjs.c (hunk 8 above)
Comment (chqrlie) [reply quickjs.c]: Actually, I found a much simple way:  if the `DUMP_XXX` is defined as nothing or 0, output is unconditional, if it is defined as a bit value, output is conditional on `rt->dump_flag & DUMP_XXX`, set via the command line option.  The `DUMP_XXX_FLAG` definitions are gone and the `CMakefile` unchanged.  Commit pending.

File: CMakeLists.txt (hunk 1 above)
Comment (chqrlie) [reply CMakeLists.txt]: This conversation is resolved as a more recent commit gives `DUMP_LEAKS` its previous behavior.

File: quickjs.c (hunk 6 above)
Comment (chqrlie) [reply quickjs.c]: Resolved by more recent commit: added `DUMP_ATOM_LEAKS` for behavior previously held by `DUMP_LEAKS=2`

File: quickjs.c (hunk 5 above)
Comment (chqrlie) [reply quickjs.c:86]: Conversation resolved with more elegant approach in latest commit

File: quickjs.h (hunk 4 above)
Comment (chqrlie) [reply quickjs.h]: Done with `uint64_t flags`

File: quickjs.c (hunk 10 above)
Comment (chqrlie) [reply quickjs.c]: More recent version makes all `DUMP_BYTECODE ` variants separate `DUMP_XXX` flags with the same semantics as the other ones.

**Comment by saghul** [comment]
This needs a rebase now that you did the split and the other stuff can be merged.

**Comment by chqrlie** [comment]
> This needs a rebase now that you did the split and the other stuff can be merged.

Yes, I am aware of this and it would have been difficult to avoid it. 
Can I just rebase my branch on master and push again?

**Comment by saghul** [comment]
Yes you can do that. Make sure to use --force then.

**Review by saghul — APPROVED** [review]

**Comment by chqrlie** [comment]
@bnoordhuis do you have any remarks? can I proceed and merge this?
